src/Spice86.Core/Emulator/CPU/ModRM.cs

```diff
@@ -93,15 +93,31 @@ public class ModRM {
             MemoryAddress = null;
             return;
         }
-        short disp = 0;
-        if (mode == 1) {
-            disp = (sbyte)_cpu.NextUint8();
-        } else if (mode == 2) {
-            disp = (short)_cpu.NextUint16();
+        
+        if (_cpu.AddressSize == 16) {
+            short displacement = mode switch {
+                1 => (sbyte)_cpu.NextUint8(),
+                2 => (short)_cpu.NextUint16(),
+                _ => 0
+            };
+            ushort offset = ComputeOffset16(mode);
+            MemoryOffset = (ushort)(offset + displacement);
+        } else {
+            long offset = ComputeOffset32(mode, _registerMemoryIndex);
```

Comment:
Why long?

Reply:
Because that saves from having to cast the result of adding the displacement back to uint (for the comparison) and then back to ushort.
I did miss a check to see if the offset becomes negative. Which I;ve now added. As well as updated the exception.